File: Linear_regression/Linear_regression.py
import numpy as np
x = np.array([1,2,3,4,5],dtype=float)
y = np.array([2,4,6,8,10],dtype=float)
w = 0
b = 0
lr = 0.01
epoch = 10000
for i in range(epoch):
    pred = w * x + b
    loss = np.mean((pred - y)**2)
    dw = (-2/len(x)) * np.sum(x * (y - pred))
    db = (-2/len(x)) * np.sum(y - pred)
    w = w - lr * dw
    b = b - lr * db

    if i % 100 == 0:
        print(f"epoch: {i}, w: {w:.4f}, b: {b:.4f}, loss: {loss:.6f}")

print(f"\nPrediction for x=7: {w*7+b:.4f} (expected 14)")

File: Linear_regression/Linear_regression.R
N <- 5
LR <- 0.01
EPOCH <- 5000

x <- c(1, 2, 3, 4, 5)
y <- c(2, 4, 6, 8, 10)

w <- 0
b <- 0

for (epoch in 0:(EPOCH - 1)) {
    # Calculate loss
    pred <- w * x + b
    loss <- sum((y - pred)^2) / N

    if (loss < 0.000001) {
        break
    }

    if (epoch %% 100 == 0) {
        cat(sprintf("epoch: %d, w: %f, b: %f, loss: %f\n", epoch, w, b, loss))
    }

    pred <- w * x + b
    dw <- sum(-2.0 / N * x * (y - pred))
    db <- sum(-2.0 / N * (y - pred))

    w <- w - LR * dw
    b <- b - LR * db
}

cat(sprintf("Final: w: %f, b: %f, loss: %f\n", w, b, loss))